Mode sorting tool

dbochkov-flexcompute opened the pull request:
Implementation of mode sorting tool as a method of `ModeSolverData` class using `FieldData.dot`. Perhaps, we can begin with feedback on the overall implementation and whether any major changes are needed?

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+        """Given an overlap matrix pair row and column entries."""
+
+        n, k = np.shape(arr)
+        assert n == k
```
want to avoid assert statements unless in tests, let's raise an error here if they are not equal instead.

tylerflex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+        self,
+        sorting: Numpy,
+        gauge_indices: Numpy,
+        track: Literal["first", "last", "central"],
```
`Literal["first", "last", "central"]` comes up a lot, so we should probably just define it once `types.py`.

dbochkov-flexcompute replied:
Good point, I actually just spotted a typo `"central"` -> `"center"` which would've been avoided doing this

tylerflex commented on tidy3d/components/mode.py:
```diff
@@ -91,6 +91,15 @@ class ModeSpec(Tidy3dBaseModel):
         "yz plane, the ``bend_axis`` is always 1 (the global z axis).",
     )
 
+    track: Literal["first", "last", "center"] = pd.Field(
```
While we haven't done this yet, I think we should move away from `Literal[str1, str2, ...]` to using enums for things like this. Could you try adjusting the implementation to make the `TrackType` an enum? you can use this as an example:
https://pydantic-docs.helpmanual.io/usage/types/#enums-and-choices


```py
# types.py
from enum import Enum

class TrackEnum(str, Enum):
    first = 'first'
    last = 'last'
    center = 'center'
```
and then we can use `TrackEnum` in the type hints and do `TrackEnum.last` instead of eg. "last"

tylerflex commented on tidy3d/components/mode.py:
```diff
@@ -91,6 +91,15 @@ class ModeSpec(Tidy3dBaseModel):
     )
 
+    track: Literal["first", "last", "center"] = pd.Field(
+        None,
```
another comment, I'm not sure about calling the field "track", maybe it is not specific enough. Maybe `track_freq`? And also, I feel instead of eg. "first" maybe "lowest"? since we're talking about frequencies?

momchil-flex replied:
I suggest setting `"central"` as default but still allowing `None` as an input. @tylerflex how would that work with enum?

tylerflex commented on tidy3d/components/data/dataset.py:
```diff
@@ -337,6 +337,15 @@ def sel_mode_index(self, mode_index: pd.NonNegativeInt) -> FieldDataset:
+
+        fields = {}
+        for field_name, data in self.field_components.items():
+            fields[field_name] = data.isel(f=[freq_index])
```
I think `freq_index` doesn't need to be in a list, could be wrong.

momchil-flex replied:
I get caught by this every now and then, basically the difference is whether the isel-ed dim will be squeezed out or not, so if you want it to stay, you put it in a list.

dbochkov-flexcompute replied:
actually, would it make sense to make it to accept any of `freq_index`, `[freq_index]`, `[freq1, freq2, ...]` which would be directly passed to `data.isel`, so there is a choice depending on what's needed? And same for `sel_mode_index`?

momchil-flex replied:
Well, now that you bring it up, I feel like we should just wrap `.isel` in the base `AbstractFieldDataset` class:

```
def isel(self, **isel_kwargs) -> AbstractFieldDataset:
    return {key: val.isel(**isel_kwargs) for key, val in self.field_components.items()}
```

We've discussed this a few times with @tylerflex. I think one argument against is that we just don't want to end up wrapping all sorts of xarray functions. But given that we are already wrapping `isel` effectively twice, once for mode index and once for frequency index, it seems like it makes more sense to just unify this.

My main objection is actually this: when you call `isel` on a `FieldData`, the returned `FieldData` will then have a mismatch in e.g. `FieldData.monitor.freqs` and `FieldData.Ex.f`. This is an issue with the current implementation of `sel_freq_index` too. How should we handle this?

1. Wrap `isel` and let the user worry about it
2. Keep the existing functions but also define `AbstractFieldData.sel_freq_index` to change the list of frequencies in `self.monitor` before calling `super().sel_freq_index`.
3. Remove the existing isel wrappers altogether and just do it manually when needed (it's basically one line of code if you write it as dict comprehension as above...)

Having written all this up I'm kinda leaning towards 3.

tylerflex reviewed (commented):
I just left a few minor comments as I saw them, but like you said maybe let's discuss at a high level before making those changes?

So the idea here (to summarize my understanding) is that we want to be able to sort the `ModeSolverData` across the `mode_index` dimension based on how similar each mode is with regards to a given mode. We use the overlap integral to determine similarity. The reference mode (I think what determines `mode_index=0` must be selected at a given frequency so we have a "track" option in both the mode spec and the overlap_sort method to let the user select either the first, middle, or highest frequency to use as reference.

I think overall this seems good, but a few comments or discussion points:
1. Why not let the user provide a frequency or index into the frequency coordinate array specifying which frequency to track? would this complicate things? seems potentially simpler but maybe slightly harder to use if ie the central frequency is unknown?
2. How is the "closeness" defined? Think we need to say more about it in the docstrings. If the mode overlap is identical but out of phase (overlap of ie -1), does this still register as a "close" match? Are the mode profiles normalized?

Question:
1. Is `overlap_sort()` meant to be a way to do this to an existing modesolverData whereas putting `track` in the `modeSpec` is more of a way to have this done when the `ModeSolver` is called?

dbochkov-flexcompute commented:
> So the idea here (to summarize my understanding) is that ...

Sounds like you got everything correctly. Just to make sure and additional clarity: say "0" is our base/track frequency, then we compare frequency "1" to "0", "2" to "1", "3" to "2, etc (not like compare everything to "0")

> Why not let the user provide a frequency or index into the frequency coordinate array specifying which frequency to track? would this complicate things? seems potentially simpler but maybe slightly harder to use if ie the central frequency is unknown? I find the "track" slightly awkward.

That crossed my mind too while implementing that we can provide frequency by its value or index. So I guess we have three options: literal, index, or value. I wonder if it makes sense to have three separate arguments, say, `freq_pos`, `freq_index`, `freq_value`, one and only one of which must be specified? However, directly transferring that into `ModeSpec` can clutter it a bit

> How is the "closeness" defined? Think we need to say more about it in the docstrings. 

This is defined to certain degree by parameter `overlap_thresh`. Specifically, before sorting modes of one frequency based on modes of another frequency I check whether current overlap values between the two frequencies are all above it (which should be quite common), if yes then we consider that the modes of the two frequencies are already matching to save some computational time, especially for large mode numbers/large mode solver sizes.
However, when actually performing sorting, I compute overlaps for all possible pairs and then match modes with the highest overlap values. So, in some sense, there is no criterion for closeness during sorting itself, the most similar modes are matched no matter what are actual overlap values. 
Finally, after sorting I check again whether new overlap values are all above `overlap_thresh` and throw a warning if not.
Does that make sense overall? I will try to make it more clear in the function description

As for the actual value of `overlap_thresh`, I don't have a strong argument, I assumed 0.7-0.9 is a good range.
 
> If the mode overlap is identical but out of phase (overlap of ie -1), does this still register as a "close" match? Are the mode profiles normalized?

I always take absolute value when computing overlap. Are there any caveats for doing that? Yes, modes profiles are normalized by their norm (sqrt(|F.F|))

> Is `overlap_sort()` meant to be a way to do this to an existing `ModeSolverData` whereas putting track in the `ModeSpec` is more of a way to have this done when the `ModeSolver` is called?

Yes, exactly. We were discussing with Momchil whether to implement sorting as a functionality of `ModeSolver` only or `ModeSolverData` to allow more flexibility and didn't find any strong arguments against the latter. Perhaps it could be useful when dealing with large numbers of modes/large mode solver sizes, that is, compute modes, inspect, and do sorting if only necessary

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+        for step, last_ind in zip([-1, 1], [-1, num_freqs]):
+            # Start with the base frequency (storing norm to avoid double work)
+            data_template = self.sel_freq_index(freq_index=f0_ind)
+            norm_template = np.sqrt(np.abs(data_template.dot(data_template)))
```
Let's add the normalization directly in the mode solver so that all the `norm` stuff here can be removed. That is to say let's make the mode solver return normalized modes.

momchil-flex commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+        if np.all(current_overlaps > overlap_thresh):
+            return np.arange(num_modes), current_overlaps
+
+        # Compute an overlap matrix, dot products of each template mode with each mode to sort
```
One thing I am wondering is if we should do this for all modes or only for the subset of modes with `current_overlap < overlap_thresh` and maybe setting a higher `overlap_thresh` by default. I understand that it's probably best to compute the full matrix but I'm just a bit worried about performance again in the case of a large number of modes in which case also it's likely that at least a few (e.g. the spurious ones) may have a low overlap.

dbochkov-flexcompute replied:
Sounds like a good idea, I think that should work. Can it happen that some two modes become degenerate at some frequencies? Maybe only in such cases it won't work perfectly

momchil-flex replied:
Yeah it can happen. Is this hard to handle either way?

dbochkov-flexcompute commented:
Addressed what we have discussed so far:
- implemented a more adaptive selection of modes to sort as suggested by Momchil, indeed, this saves even more calculation
- mode solver returns normalized data
- switching to selecting freqs and modes by hands instead of relying on `sel_mode_index` and `sel_freq_index`
- added and switched to `ModeTrackEnum` (`None` seems to be working)
- Minor changes to decriptions
- bonus: pined random seed in mode solver so that debugging is easier

Some bugs:
- I had to add .copy() [here](https://github.com/flexcompute/tidy3d/pull/562/files#diff-6df7e10c5d324ce276be07e82f01be75f3bd4cb7c0ab0ed7b3fd65711994d867R166), otherwise for some reason field data would be changed every time .dot() is called
- .dot() and .flux() don't seems to work properly with 1D monitors (?) [here](https://github.com/flexcompute/tidy3d/pull/562/files#diff-912ff90d102db41070ef42d17b5b6cf1b8690649d19ba6edfb8c91fb070b7401R81-R82). Couldn't find a quick solution, so left it as is for now.
- for some reason can get [this warning](https://github.com/flexcompute/tidy3d/pull/562/files#diff-6df7e10c5d324ce276be07e82f01be75f3bd4cb7c0ab0ed7b3fd65711994d867R506-R512) to be displayed. Am I missing something?

momchil-flex reviewed (commented):
Nice, heading toward wrapping up. I left some cosmetic comments. Regarding this:

> * I had to add .copy() [here](https://github.com/flexcompute/tidy3d/pull/562/files#diff-6df7e10c5d324ce276be07e82f01be75f3bd4cb7c0ab0ed7b3fd65711994d867R166), otherwise for some reason field data would be changed every time .dot() is called

Sounds right, I guess I thought `squeeze` returns a copy.

> * .dot() and .flux() don't seems to work properly with 1D monitors (?) [here](https://github.com/flexcompute/tidy3d/pull/562/files#diff-912ff90d102db41070ef42d17b5b6cf1b8690649d19ba6edfb8c91fb070b7401R81-R82). Couldn't find a quick solution, so left it as is for now.

I pushed a commit on your branch fixing this.

> * for some reason can get [this warning](https://github.com/flexcompute/tidy3d/pull/562/files#diff-6df7e10c5d324ce276be07e82f01be75f3bd4cb7c0ab0ed7b3fd65711994d867R506-R512) to be displayed. Am I missing something?

Do you have a test either frontend or backend where you expect it to show?

dbochkov-flexcompute commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,226 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+
+        Parameters
+        ----------
+        track : ModeTrackEnum
```
This is how it renders currently:
![1](https://user-images.githubusercontent.com/109110435/201981042-86c325b0-92b7-4306-870e-fae05da39d6d.png)

Using `:class:.ModeTrackEnum`:
![2](https://user-images.githubusercontent.com/109110435/201981127-240f6857-09fb-4841-8aa7-2393345b5ca7.png)

And the enum class itself:
![3](https://user-images.githubusercontent.com/109110435/201981170-30db79ca-b19e-4ba0-a6a0-e142329f5e83.png)

momchil-flex replied:
We want them to do `track_freq="central"` I believe. So it seems to me that in the docs we unfortunately have to manually put something like `track_freq: Literal["central", "lowest", "highest"]`. @tylerflex do you have any comment on this?

dbochkov-flexcompute commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,226 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+        data_template: ModeSolverData,
+        data_to_sort: ModeSolverData,
+        overlap_thresh: float,
+    ) -> Tuple[Numpy, Numpy]:
```
Agree, it does make more sense

dbochkov-flexcompute commented on tidy3d/components/data/monitor_data.py:
```diff
@@ -424,6 +424,226 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
+
+        return pairs, overlaps
+
+    def _find_closest_pairs(arr: Numpy) -> Tuple[Numpy, Numpy]:
```
Got it! Thanks for the overview!

momchil-flex commented on tidy3d/plugins/mode/solver.py:
```diff
@@ -525,6 +525,7 @@ def set_initial_vec(Nx, Ny, mat_dtype=np.complex128, is_tensorial=False):
 
     # Initialize the vector
     size = (Nx, Ny, len_multiplier)
+    np.random.seed(0)
```
One problem is that this will not affect random generation used internally inside numpy/scipy functions. I think this is not a problem here, because we actually have our own `set_initial_vec` function. However, if we didn't have that, and wanted to seed the in-built `scipy.linalg.eigs` initial vector generation, I don't know how to do it correctly if not with `np.random.seed(0)`.

Because of this I feel like we should stick with `np.random.seed(0)` to make sure that both our functions and internal functions work deterministically?

dbochkov-flexcompute replied:
Ah, yeah, makes sense. I looked at the source code of `scipy.linalg.eigs` out of curiosity. Turns out, it relies on fortran library arpack which internally generates initial random vector if it is not provided https://github.com/opencollab/arpack-ng/blob/master/SRC/sgetv0.f. Other than that it seems there is no other randomness in the eigenvalue solver
